File: android/app/src/main/kotlin/com/siddhant/the_router/MainChannelService.kt
package com.siddhant.the_router

import android.app.AlertDialog
import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Service class for handling main channel operations
 * Includes device info, battery level, native dialog, and vibration
 */
class MainChannelService {
  
  fun handleMethodCall(call: MethodCall, result: MethodChannel.Result, activity: MainActivity) {
    when (call.method) {
      "getDeviceInfo" -> {
        val deviceInfo = getDeviceInfo()
        result.success(deviceInfo)
      }
      "getBatteryLevel" -> {
        val batteryLevel = getBatteryLevel()
        result.success(batteryLevel)
      }
      "showNativeDialog" -> {
        val title = call.argument<String>("title") ?: "Flutter Demo"
        val message = call.argument<String>("message") ?: "This is a native dialog!"
        showNativeDialog(activity, title, message) { response ->
          result.success(response)
        }
      }
      "vibrate" -> {
        val duration = call.argument<Int>("duration") ?: 500
        val success = vibrate(activity, duration)
        result.success(success)
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  private fun getDeviceInfo(): Map<String, String> {
    return mapOf(
      "platform" to "Android",
      "version" to Build.VERSION.RELEASE,
      "device" to "${Build.MANUFACTURER} ${Build.MODEL}",
      "sdkLevel" to Build.VERSION.SDK_INT.toString(),
      "hardware" to Build.HARDWARE,
      "product" to Build.PRODUCT
    )
  }

  private fun getBatteryLevel(): Int {
    // This is a simplified implementation
    // In a real app, you'd use BatteryManager to get actual battery level
    return (Math.random() * 100).toInt()
  }

  private fun showNativeDialog(
    activity: MainActivity,
    title: String,
    message: String,
    callback: (Boolean) -> Unit
  ) {
    activity.runOnUiThread {
      AlertDialog.Builder(activity)
        .setTitle(title)
        .setMessage(message)
        .setPositiveButton("OK") { _, _ ->
          callback(true)
        }
        .setNegativeButton("Cancel") { _, _ ->
          callback(false)
        }
        .setOnCancelListener {
          callback(false)
        }
        .show()
    }
  }

  private fun vibrate(activity: MainActivity, duration: Int): Boolean {
    return try {
      val vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        val vibratorManager = activity.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
        vibratorManager.defaultVibrator
      } else {
        @Suppress("DEPRECATION")
        activity.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
      }

      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
        vibrator.vibrate(VibrationEffect.createOneShot(duration.toLong(), VibrationEffect.DEFAULT_AMPLITUDE))
      } else {
        @Suppress("DEPRECATION")
        vibrator.vibrate(duration.toLong())
      }
      true
    } catch (e: Exception) {
      false
    }
  }
}

File: ios/Runner/MainChannelService.swift
import AudioToolbox
import Flutter
import UIKit

/// Service class for handling main channel operations
/// Includes device info, battery level, native dialog, and vibration
class MainChannelService {

    func handleMethodCall(
        call: FlutterMethodCall, result: @escaping FlutterResult, activity: AppDelegate
    ) {
        switch call.method {
        case "getDeviceInfo":
            let deviceInfo = getDeviceInfo()
            result(deviceInfo)
        case "getBatteryLevel":
            let batteryLevel = getBatteryLevel()
            result(batteryLevel)
        case "showNativeDialog":
            let args = call.arguments as? [String: Any]
            let title = args?["title"] as? String ?? "Flutter Demo"
            let message = args?["message"] as? String ?? "This is a native dialog!"
            showNativeDialog(activity: activity, title: title, message: message) { response in
                result(response)
            }
        case "vibrate":
            let args = call.arguments as? [String: Any]
            let duration = args?["duration"] as? Int ?? 500
            let success = vibrate(duration: duration)
            result(success)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func getDeviceInfo() -> [String: String] {
        let device = UIDevice.current
        let systemInfo = UIDevice.current.systemName + " " + UIDevice.current.systemVersion
        let model = UIDevice.current.model
        let name = UIDevice.current.name

        return [
            "platform": "iOS",
            "version": UIDevice.current.systemVersion,
            "device": name,
            "model": model,
            "systemInfo": systemInfo,
        ]
    }

    private func getBatteryLevel() -> Int {
        UIDevice.current.isBatteryMonitoringEnabled = true
        return Int(UIDevice.current.batteryLevel * 100)
    }

    private func showNativeDialog(
        activity: AppDelegate,
        title: String,
        message: String,
        completion: @escaping (Bool) -> Void
    ) {
        DispatchQueue.main.async {
            let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)

            alert.addAction(
                UIAlertAction(title: "OK", style: .default) { _ in
                    completion(true)
                })

            alert.addAction(
                UIAlertAction(title: "Cancel", style: .cancel) { _ in
                    completion(false)
                })

            activity.window?.rootViewController?.present(alert, animated: true)
        }
    }

    private func vibrate(duration: Int) -> Bool {
        // iOS doesn't have a direct vibration API like Android
        // We use AudioToolbox for haptic feedback
        do {
            if #available(iOS 10.0, *) {
                let impactFeedback = UIImpactFeedbackGenerator(style: .medium)
                impactFeedback.impactOccurred()
            } else {
                // Fallback for older iOS versions
                AudioServicesPlaySystemSound(kSystemSoundID_Vibrate)
            }
            return true
        } catch {
            return false
        }
    }
}
